### DISCUSSION TITLE
Translation for strings in @Theia/* packages

### DISCUSSION THREAD

**Discussion by arunswarnam** [eclipse-theia/theia#11356] [Improvements]
Hi!
  I have a use case where I am interested in providing translations for all user facing strings including the ones that are coming from @theia scoped packages. Until version @theia/1.19.0 I was able to use the @theia/cli command 
`yarn extract --output ../output.json --root ../node_modules/@theia/ --logs ../log.txt --files "**/*.tsx" --files "**/*.ts"` which nicely extracted and placed all the strings into output.json but with @theia/1.26.0 I am noticing change in the behavior of the command:
1. It doesn't maintain the hierarchy of strings like for example @theia/cli gives the following output.json 
@theia : 1.19.0
```
{
    "vscode": {
        "editorOptions": {
                    "acceptSuggestionOnCommitCharacter": "Controls whether suggestions should be accepted on commit characters ..."
          }
      }
}
```
 @theia 1.26.0
```
 {
   "acceptSuggestionOnCommitCharacter": "Controls whether suggestions should be accepted on commit characters.
  }
```

2. @theia/cli misses to extract some strings in version 1.26.0 like if you notice in the attached `output.txt` strings with key [vscode/fileActions/newFolder](https://github.com/eclipse-theia/theia/blob/v1.26.0/packages/workspace/src/browser/workspace-commands.ts#L100) seem to be missing in 1.26.0.

[output_theia1.19.0.txt](https://github.com/eclipse-theia/theia/files/9002183/output_theia1.19.0.txt)
[output_theia1.26.0.txt](https://github.com/eclipse-theia/theia/files/9002193/output_theia1.26.0.txt)

This will prevent us from bringing our own translations.

**Comment by vince-fugnitto** [comment]
@arunswarnam thank you for the discussion:

1. I believe `acceptSuggestionOnCommitCharacter` is already translated by default when using vscode language-packs:

https://github.com/eclipse-theia/theia/blob/c9b9b4393e09adbf579a2ef3cd26fdc166b373c8/packages/core/src/common/i18n/nls.metadata.json#L4778

For other items (theia-specific translations that are not supported in packs) there is a hierarchy:

```json
...
"theia": {
  "callhierarchy": {
    "open": "Open Call Hierarchy"
  },
...
}
```

2, `newFolder` is translated by vscode language-packs by default (`toDefaultLocalizedCommand`):

https://github.com/eclipse-theia/theia/blob/c9b9b4393e09adbf579a2ef3cd26fdc166b373c8/packages/workspace/src/browser/workspace-commands.ts#L97-L101

https://github.com/eclipse-theia/theia/blob/c9b9b4393e09adbf579a2ef3cd26fdc166b373c8/packages/core/src/common/i18n/nls.metadata.json#L6926

**Reply by arunswarnam** [reply]
@vince-fugnitto Thank you for responding. I understand vscode language packs are capable of bringing in translations for keys that used to start with `vscode/` but we couldn't use vscode-language packs for some restrictions and we are trying to explore if we can bring in our own translations.
1. With theia version 1.26.0, I attempted to override the translation for both keys ( acceptSuggestionOnCommitCharacter and vscode/editorOptions/acceptSuggestionOnCommitCharacter) through my [custom localization provider ](https://theia-ide.org/docs/i18n). I see Theia is able to pick up translations offered for the key `vscode/editorOptions/acceptSuggestionOnCommitCharacter` similar to the [key](https://github.com/microsoft/vscode-loc/blob/release/1.64.0/i18n/vscode-language-pack-fr/translations/main.i18n.json#L247) coming from the language pack.

**Reply by vince-fugnitto** [reply]
@arunswarnam unfortunately, we currently do not support (nor plan to support any time soon) other ways to translate parts of the code from vscode without the language-packs. 

I believe that you'll need to implement and maintain this use-case for your application, and we cannot guarantee the the output  of `nls-extract` won't change going forward as we continue simplifying and refining the process.

Alternatively, you can inspire yourself from the `nls-extract` command to produce an output that is most helpful for yourself rather than use the default one directly. The functionality is implemented in [`@theia/localization-manager`](https://github.com/eclipse-theia/theia/tree/master/dev-packages/localization-manager).

---

Note: if helpful you can also take a look at [vscode-loc](https://github.com/microsoft/vscode-loc/tree/main/i18n) for the translations of vscode strings.

**Reply by arunswarnam** [reply]
I just made small edit to my previous comment. In this case all we need is to make sure the full key path (vscode/...) is logged when we extract. Let me check if there is an easier way to customize the `nls-extract` to produce the output in a desired way as suggested.
